lib/libshare/os/linux/nfs.c

```diff
@@ -578,20 +497,20 @@ nfs_enable_share(sa_share_impl_t impl_share)
 	if (error != SA_OK) {
 		unlink(filename);
 		free(filename);
-		nfs_exports_unlock();
+		nfs_exports_unlock(ZFS_EXPORTS_LOCK);
 		return (error);
 	}
 
 	error = foreach_nfs_host(impl_share, filename, nfs_add_entry,
 	    linux_opts);
 	free(linux_opts);
 	if (error == 0) {
-		error = nfs_fini_tmpfile(filename);
+		error = nfs_fini_tmpfile(ZFS_EXPORTS_FILE, filename);
 	} else {
 		unlink(filename);
 		free(filename);
 	}
-	nfs_exports_unlock();
+	nfs_exports_unlock(ZFS_EXPORTS_LOCK);
 	return (error);
```

Comment:
I think this code could be restructured to better match the freebsd implementation:

```
if (error != 0) {
    unlink(filename);
    free(filename);
    nfs_exports_unlock(ZFS_EXPORTS_LOCK);
    return (error);
}
error = nfs_fini_tmpfile(ZFS_EXPORTS_FILE, filename);
nfs_exports_unlock(ZFS_EXPORTS_LOCK);
return (error);
```